lib/experimental/class-wp-webfonts.php

```diff
@@ -14,238 +14,356 @@
+		if ( isset( $this->registered[ $font_family_handle ] ) ) {
+			return true;
+		}
+
+		$registered = $this->add( $font_family_handle, false );
+		if ( ! $registered ) {
+			return false;
+		}
+		return $this->add_data( $font_family_handle, 'is_font_family', true );
 	}
 
 	/**
-	 * Get the list of providers.
+	 * Removes a font family and all registered variations.
 	 *
-	 * @since 6.0.0
+	 * @since X.X.X
 	 *
-	 * @return WP_Webfonts_Provider[] All registered providers, each keyed by their unique ID.
+	 * @param string $font_family_handle The font family to remove.
 	 */
-	public function get_providers() {
-		return $this->providers;
+	public function remove_font_family( $font_family_handle ) {
+		if ( ! isset( $this->registered[ $font_family_handle ] ) ) {
+			return;
+		}
+
+		$variations = $this->registered[ $font_family_handle ]->deps;
+
+		foreach ( $variations as $variation ) {
+			$this->remove( $variation );
+		}
+
+		$this->remove( $font_family_handle );
 	}
 
 	/**
-	 * Register a webfont.
+	 * Add a variation to an existing family or register family if none exists.
 	 *
-	 * @since 6.0.0
+	 * @since X.X.X
 	 *
-	 * @param array $webfont Webfont to be registered.
-	 * @return string|false The font family slug if successfully registered, else false.
+	 * @param string $font_family_handle The font family's handle for this variation.
+	 * @param array  $variation          An array of variation properties to add.
+	 * @param string $variation_handle   Optional. The variation's handle. When none is provided, the
+	 *                                   handle will be dynamically generated.
+	 *                                   Default empty string.
+	 * @return string|null Variation handle on success. Else null.
 	 */
-	public function register_webfont( array $webfont ) {
-		$webfont = $this->validate_webfont( $webfont );
+	public function add_variation( $font_family_handle, array $variation, $variation_handle = '' ) {
+		if ( ! WP_Webfonts_Utils::is_defined( $font_family_handle ) ) {
+			trigger_error( 'Font family handle must be a non-empty string.' );
+			return null;
+		}
 
-		// If not valid, bail out.
-		if ( ! $webfont ) {
-			return false;
+		if ( '' !== $variation_handle && ! WP_Webfonts_Utils::is_defined( $variation_handle ) ) {
+			trigger_error( 'Variant handle must be a non-empty string.' );
+			return null;
+		}
+
+		// Register the font family when it does not yet exist.
+		if ( ! isset( $this->registered[ $font_family_handle ] ) ) {
+			if ( ! $this->add_font_family( $font_family_handle ) ) {
+				return null;
+			}
+		}
+
+		$variation = $this->validate_variation( $variation );
+
+		// Variation validation failed.
+		if ( ! $variation ) {
+			return null;
 		}
 
-		$slug = $this->get_font_slug( $webfont );
+		if ( '' === $variation_handle ) {
+			$variation_handle = WP_Webfonts_Utils::convert_variation_into_handle( $font_family_handle, $variation );
+			if ( is_null( $variation_handle ) ) {
+				return null;
+			}
+		}
```

Comment:
Hmm, that line of above it has been changed in https://github.com/WordPress/gutenberg/pull/43492/commits/f497a97cb458a1afe84cbc1d5ee4b39a25a71fd4. Previously that line of code was this:

```php
if ( '' !== $variation_handle && ! WP_Webfonts_Utils::is_defined( $variation_handle ) ) {
```

meaning it only checked if the handle had a value but was not an empty string. Why? I have to back to review, but I think it was because the `theme.json` structure is changed with this new rearchitectural redesign. The code was allowing for the old and new structure to work. Will reverify.